[PK-211] Вывод кода сохранения CelestaSQL из класса Score и модификация

packpaul opened the pull request:
dbschemasync
implementaion of GrainElement.save() extracted to CelestaSerializer

## Overview

Please describe your changes here.

---

### Checklist

- [x] Change is covered by automated tests.
- [x] JavaDoc is provided for all new/changed public APIs.
- [ ] User Guide is updated.
- [x] CI builds pass.

packpaul committed:
[PK-211] Extraction of code of saving to CelestaSQL from Score class and
dbschemasync modification
implementaion of GrainElement.save() extracted to CelestaSerializer

inponomarev commented on celesta-sql/src/main/java/ru/curs/celesta/score/MaterializedView.java:
```diff
@@ -304,7 +294,7 @@ void addFromTableRef(TableRef ref) throws ParseException {
 
     public String getChecksum() {
         try (ChecksumInputStream is = new ChecksumInputStream(
```
@packpaul: Unwanted coupling here: `CelestaSerializer` was not intended to be used from GrainElement classes. 

I looked at the code and I see no easy way to get around this problem, so maybe this place will be a technical debt.

inponomarev replied:
**This is still an unwanted coupling.** Maybe a code comment should be added here, something like 
```java
//TODO: CelestaSerializer is not intended to be used from GrainElement classes.
```

packpaul replied:
I'v got no other ideas on how to calculate the checksum, so I'll put a TODO: clause.

inponomarev replied:
👌

inponomarev commented on celesta-sql/src/main/java/ru/curs/celesta/score/CelestaSerializer.java:
```diff
@@ -0,0 +1,563 @@
+package ru.curs.celesta.score;
```
Nice, after all the `save` methods moved to this class, this started to resemble "Visitor" pattern. This can be further refactored, but I like it already!

packpaul commented:
Ivan please keep calm this is still a draft PR.

inponomarev commented:
I just wanted to give you an early feedback 😃

packpaul committed:
[PK-211] Extraction of code of saving to CelestaSQL from Score class and
dbschemasync modification
Namespace concept introduced; score saving extracted to GrainSaver;
dbschema use cases expanded

packpaul committed:
[PK-211] Extraction of code of saving to CelestaSQL from Score class and
dbschemasync modification
trailing spaces clean-up

packpaul committed:
[PK-211] Extraction of code of saving to CelestaSQL from Score class and
Unit test fixes

packpaul committed:
[PK-211] Extraction of code of saving to CelestaSQL from Score class and
Satisfying findbugs

inponomarev commented on celesta-sql/src/main/java/ru/curs/celesta/score/discovery/ScoreByScorePathDiscovery.java:
```diff
@@ -49,15 +50,29 @@ public ScoreByScorePathDiscovery(String scorePath) {
-                    .map(Path::toFile)
-                    .map(FileResource::new)
+                    .map(p -> {
+                        return new FileResource(p.toFile(), getNamespaceFromPath(scorePath.relativize(p)));
```
It will be a bit more clear if you use expression lambda
```java
p -> new FileResource(...)
```
without braces & `return`

packpaul replied:
Done.

inponomarev commented on celesta-sql/src/main/java/ru/curs/celesta/score/discovery/ScoreByScorePathDiscovery.java:
```diff
@@ -49,15 +50,29 @@ public ScoreByScorePathDiscovery(String scorePath) {
         }
     }
 
+    Namespace getNamespaceFromPath(Path p) {
```
1. I generally dislike the idea of returning `null`. What does `null Namespace` mean? Maybe we could devise a special `DefaultNamespace` for this case? (However, I not insisting.)

2. I love `Collectors.joining`. How about
```java
 IntStream.range(0, p.getNameCount() - 1)
                .mapToObj(p::getName)
                .map(Path::toString)
                .map(String::toLowerCase)
                .collect(Collectors.joining("."));
```
?

packpaul replied:
2. Done

packpaul replied:
1. Introduced Namespace.DEFAULT

inponomarev commented on celesta-sql/src/main/java/ru/curs/celesta/score/discovery/ScoreByScoreResourceDiscovery.java:
```diff
@@ -89,4 +90,19 @@ String getGrainName(String grainPath) {
+    Namespace getGrainNamespace(String grainPath) {
+
+        String value = grainPath.toLowerCase();
+
```
1. The same concern about `null` Namespace.
2. BTW this can be rewritten in a way that resembles `ScoreByScorePathDiscovery`

```java
String[] parts = grainPath.split("/");
IntStream.range(0, parts.length-1)
                .mapToObj(i->parts[i])
                .map(String::toLowerCase)
                .collect(Collectors.joining("."));
```
(proposing this just for fun, not insisting)

packpaul replied:
2. Done. You're right. We need to keep up with the time and use all that surgar features of the Java langauage. I justify myself only for the reason of saving my time since I was tired (lazy ?) to perform all that mental streams gymnastics.

inponomarev replied:
Well generally I don't like using streams for streams' sake. 

However,  in this particular case, streams give us an opportunity to write two idiomatically similar `getGrainNamespace`  methods in a way that their code also looks similar, so why not using streams?

inponomarev commented on celesta-sql/src/main/java/ru/curs/celesta/score/Grain.java:
```diff
@@ -547,8 +459,36 @@ void addNativeSql(String sql, boolean isBefore, DBType dbType, GrainPart grainPa
     }
 
-    public Set<GrainPart> getGrainParts() {
+    public Set<? extends GrainPart> getGrainParts() {
```
What was the problem here? `grainParts` has ` Set<GrainPart>` type, returning wildcard types should be avoided. Do we really need this?

packpaul replied:
Yes, the goal here was to get rid of the **set** modification over getter like:
`grain.getGrainParts().add(...)`, since changing collections in this way may be fragile (the incapsulated collection could become f.e. unmodifiable in the future). Therefore `addGrainPart()` was excplicitly added to the Grain API. I'll remove the wilcard though.

inponomarev replied:
Wow, I understood the idea! Never thought about making collections immutable this way. 🤔  This, however, will not prevent anyone from calling `Set.remove(Object o)`. 

So... either `Collections.immutableSet` or just let's give it up and let's suppose that misusing this getter is the misusers' problem.

packpaul replied:
I did change back to `public Set<GrainPart> getGrainParts()`

inponomarev commented on celesta-sql/src/main/java/ru/curs/celesta/score/GrainPart.java:
```diff
@@ -37,6 +37,20 @@ Grain getGrain() {
+     *
+     * @return
+     */
+    public Namespace getNamespace() {
```
see my another comments about `null` Namespace.

packpaul committed:
[PK-211] Extraction of code of saving to CelestaSQL from Score class and
Adjustments after code review

inponomarev commented on celesta-sql/src/main/java/ru/curs/celesta/score/discovery/ScoreByScorePathDiscovery.java:
```diff
@@ -53,26 +54,21 @@ public ScoreByScorePathDiscovery(String scorePath) {
-        }
-
-        return new Namespace(sb.toString().toLowerCase());
+        return IntStream.range(0, p.getNameCount() - 1)
```
I see you removed the `if (p.getNameCount() <= 1)`. Is there a unit case for this case?

packpaul replied:
added

inponomarev commented on celesta-sql/src/main/java/ru/curs/celesta/score/Grain.java:
```diff
@@ -473,14 +473,14 @@ public Namespace getNamespace() {
         while (i.hasNext()) {
-            if ((ns == null) || !ns.equals(i.next().getNamespace())) {
-                return null;
+            if (ns.equals(Namespace.DEFAULT) || !ns.equals(i.next().getNamespace())) {
```
`ns.equals(Namespace.DEFAULT)` is generally not null-safe, it's always better to write `Namespace.DEFAULT.equals(ns)` or even `ns == Namespace.DEFAULT`.

packpaul replied:
Done: `ns.equals(Namespace.DEFAULT)` -> `Namespace.DEFAULT.equals(ns)`

packpaul committed:
[PK-211] Extraction of code of saving to CelestaSQL from Score class and
Adjustments after code review #2

inponomarev commented:
👍
